design/one-pager-automatic-dependency-downgrades.md

```diff
@@ -0,0 +1,76 @@
+quick setup but could not handle dynamic environments where dependency version change is required. In cases where the 
+dependency was installed but the constraints were no longer satisfied, the lock reconciler gives no error and package 
+revision reconciler notices the invalid version issue and returns an error to the user for manual intervention.
+
+With recent updates, the package manager has gained the capability to upgrade dependencies to the minimum valid version 
+rather than the latest, reducing the likelihood of breaking changes in existing environments. If no version meets the 
+constraints, the lock reconciler now displays the error message in the lock object’s status, complementing the error 
+messages from the package revision reconciler.
+
+It was deliberately decided not to downgrade the dependencies automatically to prevent unexpected behaviours in the 
+existing environments. However, this approach has its own drawbacks. The user has to manually intervene to downgrade the
+dependencies, which is not a good user experience and is not suitable for GitOps workflows. This proposal aims to 
+address this issue by introducing automatic dependency version downgrades behind a feature flag.
+
+![Package Manager diagram][pkg-mgr-diagram]
+
+## Possible Issues
+
+During dependency version downgrades, the following issues may occur:
+
+1) Missing CRDs
+2) Missing CRD fields
+3) CRD storage version changes
+
+Please note that these issues may also occur during package upgrades. But in that case, we assume the user is aware of 
+the changes and accepts the risks.
+
+### Missing CRDs
+
+The CRDs installed with the newer version may not exist in the older one. This can result in custom resources 
+without associated controllers, potentially causing unexpected behavior, such as lack of updates on the cloud 
+provider's side after updating the MR. Updating the parent package to upgrade its dependencies to the newer version
+resolves this issue.
+
+### Missing CRD Fields
+
+CRD fields may change between versions. If the newer version introduces additional fields, the older version will remove
+them. Downgrading the package can lead to the loss of these values, and upgrading back to the newer version 
+cannot restore them to their previously assigned states.
+ 
+### CRD Storage Version Changes
+
+CRD storage versions may change between versions. If the storage version is upgraded in the newer version, trying to 
+downgrade the package will not automatically change the storage version on Kubernetes and fail. Providers in Upbound 
+Official Program do not change the storage version between minor versions to prevent this issue. However, this is not
+guaranteed for other providers. 
+
+## Proposal
+
+The automatic dependency version downgrade feature will be implemented behind a feature flag. The user can enable this
+feature by setting the `AutomaticDependencyDowngrade` environment variable to `true` in the package manager configuration. 
```

Comment:
It would be good to focus on helm API which is what users would interact rather then passing env vars directly to the package manager. Also, thanks to [the kong library](https://github.com/alecthomas/kong), we support both env var and argument for [any flag](https://github.com/crossplane/crossplane/blob/528a750776b0ab1e1aafaf68e3884bea23281f8c/cmd/crossplane/core/core.go#L88-L134) we add.